# python-tutorial
### Introduction

#### Definition
Python is a dynamic and strongly typed programming language. It employs both duck typing and gradual typing, via type hints.

Imperative, declarative (e.g., functional), and object-oriented programming styles are all supported, but internally everything in Python is an object.

Python puts a strong emphasis on code readability and (similar to Haskell) uses significant indentation to denote function, method, and class definitions.

Python was created by Guido van Rossum and first released in 1991.

#### Why Python?

1.	Readability: Python’s syntax is designed to be clear and straightforward. It reads almost like plain English, making it easy for newcomers to understand and write code.
2.	Versatility: Python is a Swiss Army knife of programming languages. You can use it for web development, data analysis, scientific computing, automation, artificial intelligence, and more.
3.	Large Community: Python boasts a vibrant and supportive community. There are countless tutorials, libraries, and forums to help you learn and solve problems.
4.	Cross-Platform: Whether you use Windows, macOS, or Linux, Python works seamlessly on all major operating systems.
5.	Rich Ecosystem: Python comes with a vast standard library that provides pre-built modules for various tasks, saving 
	you time and effort.




### Comment 
We can add comments to our Python scripts. Comments are important to make sure that you and others can understand what your code is about and do not run as Python code. They start with # tag. 







### Python as a calculator
Python is perfectly suited to do basic calculations. It can do addition, subtraction, multiplication and division.

In [22]:
# Addition
print(4+5)

# Subtraction
print(5 - 5)

# Multiplication
print(3*5)

# Division
print(10/2)


9
0
15
5.0


### Variables 

 A variable is a name that refers to a value stored in the computer's memory. It acts as a placeholder or a label, allowing programmers to access and manipulate data throughout their code. In Python, a variable allows you to refer to a value with a name. Variables are created by assigning a value to them using the equal sign (=), and their values can be changed as needed.
 

Python is dynamically typed, meaning we don't need to declare a variable's type ahead of time; the type is inferred at runtime based on the assigned value.

In [23]:
x = 10        # An integer assignment
name = "Leila" # A string assignment

# Here, x and name are variables. x holds an integer value 10, and name holds the string "Leila".

print(x)
name

10


'Leila'

In [24]:
# Create a variable savings
savings = 100

# Print out savings
print(savings)

100


#### Calculations with variables
You've now created a savings variable, so let's start saving!

Instead of calculating with the actual values, you can use variables instead. The savings variable you created in the previous exercise with a value of 100 is available to you.

How much money would you have saved four months from now, if you saved £10 each month?

In [25]:
# Create the variables monthly_savings and num_months
monthly_savings = 10
num_months = 4

# Multiply monthly_savings and num_months
new_savings = monthly_savings * num_months

# Add new_savings to your savings
total_savings = savings + new_savings

# Print total_savings
print(total_savings)


140


### Variables Type (Data Types in Python)

1. integers (int): Used for whole numbers.
2. Floating-point numbers (float): Used for numbers with decimals or in exponential form.
3. Strings (str): Used for textual data. You can use single or double quotes to build a string.
4. Booleans (bool): a type to represent logical values. It can only be True or False (the capitalization is important!).
5. Lists (list): Mutable sequences used to store collections of items (elements can be added, removed, or changed).
6. Tuples (tuple): Immutable sequences, meaning once created, their elements cannot be changed.
7. Dictionaries (dict): Mutable collections that store data as key-value pairs.
8. Sets (set): Mutable collections of unique elements used to perform common set operations.
9. Each of these types plays a crucial role in handling data in Python, enabling developers to solve a wide 

In [26]:
# Create a variable half
half = 0.5

# Create a variable intro
intro = "Hello! How are you?"

# Create a variable is_good
is_good = True 

#### Guess the type
To find out the type of a value or a variable that refers to that value, you can use the type() function. Suppose you've defined a variable a, but you forgot the type of this variable. To determine the type of a, simply execut: type(a)



In [27]:
type(half)

float

#### Operations with other types
Note: Different types behave differently in Python.

When you sum two strings, for example, you'll get different behavior than when you sum two integers or two booleans.


In [28]:
monthly_savings = 10
num_months = 12
intro = "Hello! How are you?"

# Calculate year_savings using monthly_savings and num_months
year_savings = monthly_savings * num_months

# Print the type of year_savings
print(type(year_savings))

# Assign sum of intro and intro to doubleintro
doubleintro = intro + intro

# Print out doubleintro
print(doubleintro)

<class 'int'>
Hello! How are you?Hello! How are you?


#### Type conversion
Using the + operator to paste together two strings can be very useful in building custom messages.

Suppose, for example, that you've calculated your savings want to summarize the results in a string.

To do this, you'll need to explicitly convert the types of your variables. More specifically, you'll need str(), to convert a value into a string. str(savings), for example, will convert the integer savings to a string.

Similar functions such as int(), float() and bool() will help you convert Python values into any type.

In [29]:
# Definition of savings and total_savings
savings = 100
total_savings = 150

# Fix the printout
print("I started with $" + str(savings) + " and now have $" + str(total_savings) + ". Awesome!")

# Definition of pi_string
pi_string = "3.1415926"

# Convert pi_string into float: pi_float
pi_float = float(pi_string)
pi_float

I started with $100 and now have $150. Awesome!


3.1415926

##### Arithmetic operations with different types 
Note : In Python, the boolean values True and False are treated as 1 and 0 respectively when used in arithmetic contexts. Therefore, adding True and False together results in 1 (since 1 + 0 = 1)

In [30]:
"I can add integers, like " + str(5) + " to strings."

"I said " + ("Hey " * 2) + "Hey!"

"The correct answer to this multiple choice exercise is answer number " + str(2)

True + False

1

### List
As opposed to int, bool etc., a list is a compound data type; you can group values together.

- Data Structure: A list is an ordered collection of items, which can be of different types (integers, strings, other lists, etc.).
- Mutable: Lists can be modified after their creation; items can be added, removed, or changed.
- Indexable: Each item in a list has a specific order and can be accessed using an index, starting from 0 for the first element.
- Flexible: Lists can grow or shrink dynamically as items are added or removed.
- Syntax: Defined by square brackets [], with items separated by commas, e.g., [1, 'hello', 3.14].
- Versatile Functions: Supports numerous methods like append(), remove(), sort(), and many more to manipulate the data.
- Supports Nesting: Lists can contain other lists, creating matrix-like structures or more complex data formats.

In [31]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Create list areas
areas = [hall,kit,liv,bed,bath]
# Print areas
print(areas)

[11.25, 18.0, 20.0, 10.75, 9.5]


##### Create list with different types
A list can contain any Python type. Although it's not really common, a list can also contain a mix of Python types including strings, floats, booleans, etc.

The printout of the previous exercise wasn't really satisfying. It's just a list of numbers representing the areas, but you can't tell which area corresponds to which part of your house.

The code in the editor is the start of a solution. For some of the areas, the name of the corresponding room is already placed in front. Pay attention here! "bathroom" is a string, while bath is a variable that represents the float 9.50 you specified earlier.

In [32]:
# area variables (in square meters)
hall = 11.25
kit = 18.0
liv = 20.0
bed = 10.75
bath = 9.50

# Adapt list areas
areas = ["hallway", hall, "kitchen", kit, "living room", liv, "bedroom", bed, "bathroom", bath]

# Print areas
print(areas)


['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0, 'bedroom', 10.75, 'bathroom', 9.5]


A list can contain any Python type. But a list itself is also a Python type. That means that a list can also contain a list!

In [33]:
A= [1, 3, 4, 2] 
B=[[1, 2, 3], [4, 5, 7]] 
C= [1 + 2, "a" * 5, 3]

C

[3, 'aaaaa', 3]

#### List of lists
You'll often be dealing with a lot of data, and it will make sense to group some of this data.

Instead of creating a flat list containing strings and floats, representing the names and areas of the rooms in your house, you can create a list of lists. 

Don't get confused here: "hallway" is a string, while hall is a variable that represents the float 11.25 you specified earlier.

In [34]:
# house information as list of lists
house = [["hallway", hall],
         ["kitchen", kit],
         ["living room", liv],
         ["bedroom", bed],
         ["bathroom", bath]]

# Print out house
print(house)

# Print out the type of house
print(type(house))


[['hallway', 11.25], ['kitchen', 18.0], ['living room', 20.0], ['bedroom', 10.75], ['bathroom', 9.5]]
<class 'list'>


#### Subset and conquer
Subsetting Python lists is a piece of cake. Take the code sample below, which creates a list x and then selects "b" from it. Remember that this is the second element, so it has index 1. You can also use negative indexing.




In [35]:
x = ["a", "b", "c", "d"]
x[1]
x[-3] #same result!

'b'

So the areas list from before, containing both strings and floats. Let's add the correct code to do some Python subsetting.

In [36]:
# Print out second element from areas
print(areas[1])

# Print out last element from areas
print(areas[-1])

# Print out the area of the living room
print(areas[5])

11.25
9.5
20.0


#### Subset and calculate
After you've extracted values from a list, you can use them to perform additional calculations. Take this example, where the second and fourth element of a list x are extracted. The strings that result are pasted together using the + operator:

In [37]:

print(x[1] + x[3])

bd


In [38]:
# Sum of kitchen and bedroom area: eat_sleep_area
eat_sleep_area = areas[3] + areas[-3]

# Print the variable eat_sleep_area
print(eat_sleep_area)


28.75


#### Slicing and dicing
Selecting single values from a list is just one part of the story. It's also possible to slice your list, which means selecting multiple elements from your list. Use the following syntax:

my_list[start:end]

The start index will be included, while the end index is not.

However, it's also possible not to specify these indexes. If you don't specify the begin index, Python figures out that you want to start your slice at the beginning of your list. If you don't specify the end index, the slice will go all the way to the last element of your list. 


In [39]:
x[1:3]
x[:2]
x[2:]
x[:]

['a', 'b', 'c', 'd']

In [40]:
# Create the areas list
areas = ["hallway", 11.25, "kitchen", 18.0, "living room", 20.0, "bedroom", 10.75, "bathroom", 9.50]

# Use slicing to create downstairs (first 6 elements)
downstairs =areas [:6]

# Use slicing to create upstairs (last 4 elements)
upstairs =areas [6:]

# Print out downstairs and upstairs
print(downstairs)
print(upstairs)

['hallway', 11.25, 'kitchen', 18.0, 'living room', 20.0]
['bedroom', 10.75, 'bathroom', 9.5]
